File: soungmunkim/C/Graph/210_CourseSchedule2.c
/*
[directed graph 순서 나열하기 (topological order)]

There are a total of numCourses courses you have to take, 
    labeled from 0 to numCourses - 1. 
You are given an array prerequisites 
    where prerequisites[i] = [ai, bi] indicates that 
    you must take course bi first if you want to take course ai.

For example, the pair [0, 1], indicates that to take course 0 
you have to first take course 1. (directed graph)
Return the ordering of courses you should take to finish all courses. 
If there are many valid answers, return any of them. 
If it is impossible to finish all courses, return an empty array.


Example 1:
Input: numCourses = 2, prerequisites = [[1,0]]
Output: [0,1]
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0. So the correct course order is [0,1].

numCourses = 4
prerequisites = [[1,0],[2,0],[3,1],[3,2]]
findOrder(numCourses, prerequisites)
[0, 1, 2, 3]

numCourses = 1
prerequisites = []
findOrder(numCourses, prerequisites)
[0]
*/

#include <stdio.h>
#include <stdlib.h>

// 그래프 구조체 정의
typedef struct {
    int** edges;       // 엣지 정보 저장
    int* size;         // 각 노드의 엣지 개수 저장
    int numVertices;   // 그래프의 노드(정점) 개수
} Graph;

// 큐 구조체 정의
typedef struct {
    int* data;         // 데이터 저장
    int front, back;   // 큐의 앞, 뒤 위치 정보
    int capacity;      // 큐의 최대 용량
} Queue;

// 그래프 생성 함수
Graph* createGraph(int numVertices) {
    // Graph 구조체를 위한 메모리 동적 할당
    Graph* graph = (Graph*)malloc(sizeof(Graph));
    
    // 각 노드가 가지고 있는 엣지들의 목록을 위한 메모리 할당
    graph->edges = (int**)malloc(numVertices * sizeof(int*));

    // 각 노드의 엣지 개수를 저장하기 위한 메모리 할당 및 초기화
    graph->size = (int*)calloc(numVertices, sizeof(int));

    // 그래프의 전체 노드(정점) 개수 설정
    graph->numVertices = numVertices;

    // 각 노드 별로 엣지 리스트 메모리 동적 할당
    for (int i = 0; i < numVertices; i++) {
        graph->edges[i] = (int*)malloc(numVertices * sizeof(int));
    }

    // 생성된 그래프의 주소 반환
    return graph;
}


// 그래프에 엣지 추가 함수
void addEdge(Graph* graph, int src, int dest) {
    graph->edges[src][graph->size[src]++] = dest;
}

// 그래프 메모리 해제 함수
void freeGraph(Graph* graph) {
    for (int i = 0; i < graph->numVertices; i++) {
        free(graph->edges[i]);
    }
    free(graph->edges);
    free(graph->size);
    free(graph);
}

// 큐 생성 함수
Queue* createQueue(int capacity) {
    Queue* queue = (Queue*)malloc(sizeof(Queue));
    queue->data = (int*)malloc(capacity * sizeof(int));
    queue->front = 0;
    queue->back = 0;
    queue->capacity = capacity;
    return queue;
}

// 큐에 데이터 삽입 함수
void enqueue(Queue* queue, int item) {
    queue->data[queue->back++] = item;
}

// 큐에서 데이터 추출 함수
int dequeue(Queue* queue) {
    return queue->data[queue->front++];
}

// 큐가 비어있는지 확인 함수
int isEmpty(Queue* queue) {
    return queue->front == queue->back;
}

// 큐 메모리 해제 함수
void freeQueue(Queue* queue) {
    free(queue->data);
    free(queue);
}

// 주어진 조건에 맞는 순서 찾기 함수
int* findOrder(int numCourses, int** prerequisites, int prerequisitesSize, int* returnSize) {
    // 그래프 생성
    Graph* graph = createGraph(numCourses); 

    // prerequisites 정보를 바탕으로 그래프에 엣지 추가
    for (int i = 0; i < prerequisitesSize; i++) {
        addEdge(graph, prerequisites[i][1], prerequisites[i][0]);
    }

    // 각 노드의 진입차수를 저장하기 위한 배열 초기화
    int* indegree = (int*)calloc(numCourses, sizeof(int));

    // 그래프를 순회하면서 각 노드의 진입차수 계산
    for (int i = 0; i < numCourses; i++) {
        for (int j = 0; j < graph->size[i]; j++) {
            indegree[graph->edges[i][j]]++;
        }
    }

    // 진입차수가 0인 노드를 저장하기 위한 큐 생성
    Queue* queue = createQueue(numCourses);
    for (int i = 0; i < numCourses; i++) {
        if (indegree[i] == 0) {
            enqueue(queue, i);
        }
    }

    // 위상 정렬 결과를 저장할 배열 할당
    int* order = (int*)malloc(numCourses * sizeof(int));
    *returnSize = 0;

    // 큐가 빌 때까지 반복
    while (!isEmpty(queue)) {
        // 큐에서 노드를 하나 꺼냄
        int cur = dequeue(queue);
        order[(*returnSize)++] = cur;  // 결과 배열에 현재 노드 추가

        // 현재 노드와 인접한 노드의 진입차수 감소
        for (int j = 0; j < graph->size[cur]; j++) {
            int adj = graph->edges[cur][j];
            indegree[adj]--;
            // 진입차수가 0이 된 노드를 큐에 추가
            if (indegree[adj] == 0) {
                enqueue(queue, adj);
            }
        }
    }

    // 모든 정점을 방문하지 않은 경우 (사이클 존재)
    if (*returnSize != numCourses) {
        free(order);   // 결과 배열 메모리 해제
        order = NULL;  // 포인터 초기화
        *returnSize = 0;  // 반환할 배열의 크기 초기화
    }

    // 동적 할당한 메모리 해제
    freeGraph(graph);
    freeQueue(queue);
    free(indegree);

    // 최종 결과 반환
    return order;
}

// 주어진 크기와 데이터를 기반으로 방향성 있는 간선들을 생성하는 함수
int** createDirectedEdges(int size, int data[][2]) {
    // 크기에 따라 int 포인터 배열을 동적 할당합니다.
    int** edges = (int**)malloc(size * sizeof(int*));
    for (int i = 0; i < size; i++) {
        // 각 원소에 대하여 int 배열 크기 2를 할당합니다.
        edges[i] = (int*)malloc(2 * sizeof(int));
        // 데이터에서 간선의 시작점을 복사합니다.
        edges[i][0] = data[i][0];
        // 데이터에서 간선의 끝점을 복사합니다.
        edges[i][1] = data[i][1];
    }
    // 생성된 간선 리스트를 반환합니다.
    return edges;
}

// 방향성 있는 간선들의 메모리를 해제하는 함수
void freeDirectedEdges(int** edges, int size) {
    for (int i = 0; i < size; i++) {
        // 각 원소에 할당된 메모리를 해제합니다.
        free(edges[i]);
    }
    // 전체 배열에 할당된 메모리를 해제합니다.
    free(edges);
}
// 출력 함수 숫자 리스트 [1,2,3,4] 형태로 
void printOrder(int* order, int size) {
    printf("[");
    for (int i = 0; i < size; i++) {
        printf("%d", order[i]);
        if (i < size - 1) printf(",");
    }
    printf("]\n");
}


int main() {
    // 예제 1
    int prerequisitesData1[][2] = {{1, 0}};
    int** prerequisites1 = createDirectedEdges(1, prerequisitesData1);

    int returnSize1;
    int* order1 = findOrder(2, prerequisites1, 1, &returnSize1);

    // 출력 
    printOrder(order1, returnSize1);

    
    freeDirectedEdges(prerequisites1, 1);
    free(order1);

    // 예제 2
    int prerequisitesData2[][2] = {{1, 0}, {2, 0}, {3, 1}, {3, 2}};
    int** prerequisites2 = createDirectedEdges(4, prerequisitesData2);

    int returnSize2;
    int* order2 = findOrder(4, prerequisites2, 4, &returnSize2);

    // 출력
    printOrder(order2, returnSize2);

    freeDirectedEdges(prerequisites2, 4);
    free(order2);

    // 예제 3
    int returnSize3;
    int* order3 = findOrder(1, NULL, 0, &returnSize3);

    // 출력
    printOrder(order3, returnSize3);

    free(order3);

    return 0;
}

File: soungmunkim/Python/Graph/210_CourseSchedule2.py
"""
[directed graph 순서 나열하기 (topological order)]

There are a total of numCourses courses you have to take, 
    labeled from 0 to numCourses - 1. 
You are given an array prerequisites 
    where prerequisites[i] = [ai, bi] indicates that 
    you must take course bi first if you want to take course ai.

For example, the pair [0, 1], indicates that to take course 0 
you have to first take course 1. (directed graph)
Return the ordering of courses you should take to finish all courses. 
If there are many valid answers, return any of them. 
If it is impossible to finish all courses, return an empty array.


Example 1:
Input: numCourses = 2, prerequisites = [[1,0]]
Output: [0,1]
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0. So the correct course order is [0,1].

numCourses = 4
prerequisites = [[1,0],[2,0],[3,1],[3,2]]
findOrder(numCourses, prerequisites)
[0, 1, 2, 3]

numCourses = 1
prerequisites = []
findOrder(numCourses, prerequisites)
[0]
"""

from typing import List
def findOrder(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    G = dict()
    # number of courses만큼 graph adj list 만들기
    for i in range(numCourses):
        G[i] = []
        
    # 각 edge들 정보 추가하기 (directed)
    for i in range(len(prerequisites)):
        want = prerequisites[i][0]
        pre = prerequisites[i][1]
        
        G[pre].append(want)
        
    # topological sort 
    indegree = [0] * len(G) # 모든 노드의 진입차수를 0으로 초기화
    
    # 각 노드의 진입 차수 계산함
    """
    indegree 리스트는 각 노드의 초기 진입 차수를 저장하는 데 사용
    prerequisites = [[1,0],[2,0],[3,1],[3,2]]
    numCourses = 4

    이를 그래프로 나타내면 다음과 같습니다:
    0 -> 1
    0 -> 2
    1 -> 3
    2 -> 3

    따라서 진입 차수는 다음과 같습니다.
    0: 0
    1: 1
    2: 1
    3: 2
    """
    # 얼마나 들어오는 edges가 많은지
    for key in G:
        for neighbor in G[key]:
            indegree[neighbor] += 1
            
    Q = []
    for i in range(len(G)):
        if indegree[i] == 0:  # 진입차수가 0인 node를 큐에 넣는다.
            Q.append(i)
    
    order = [] # 위상 순서를 저장할 리스트

    while Q: # 큐가 비어있지 않으면 반복
        cur = Q.pop(0)
        order.append(cur) # 현재 node 리스트에 추가
        
        for adj in G[cur]: # 인접 노드 돌기
            indegree[adj] -= 1 # 인접 노드의 진입 차수 감소
            if indegree[adj] == 0: # 진입차수가 0이면 Q 추가
                Q.append(adj)
        

    # 모든 정점을 방문하지 않은 경우 (cycle 존재함)
    # 사이클을 형성하는 노드들 중 
    # 어느 노드도 진입 차수가 0이 되지 않기 때문에 큐에 들어가지 않게 됩니다.
    if len(order) != len(G):
        return []
    else:
        return order
    

## ------------------ GNode class 버전으로 구현 -------------------- ##
# from typing import List

# class GNode:
#     def __init__(self, id):
#         self.id = id                 # 노드의 ID 혹은 번호
#         self.out_nodes = []          # 이 노드로부터 나가는 간선의 목록
#         self.in_degree = 0           # 이 노드로 들어오는 간선의 개수 (진입차수)

# def findOrder(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
#     # 각 번호를 가진 GNode 객체로 그래프 초기화
#     graph = {i: GNode(i) for i in range(numCourses)}

#     # 그래프 구성하기
#     for pair in prerequisites:
#         next_course, prev_course = pair
#         graph[prev_course].out_nodes.append(next_course)   # 나가는 간선 정보 추가
#         graph[next_course].in_degree += 1                  # 진입차수 증가

#     # 진입차수가 0인 노드들로 큐 초기화
#     queue = [k for k in graph if graph[k].in_degree == 0]

#     order = []  # 위상 정렬 결과를 저장할 리스트

#     # 큐를 이용한 위상 정렬 수행
#     while queue:
#         node = queue.pop(0)  # 큐의 앞에서 노드 추출
#         order.append(node)   # 결과 리스트에 노드 추가

#         # 현재 노드의 나가는 간선에 연결된 노드들의 진입차수 감소
#         for next_node in graph[node].out_nodes:
#             graph[next_node].in_degree -= 1
#             # 진입차수가 0이 된 노드는 큐에 추가
#             if graph[next_node].in_degree == 0:
#                 queue.append(next_node)

#     # 모든 과목을 수강할 수 있는지 확인. 만약 가능하다면 위상 정렬 결과 반환, 그렇지 않다면 빈 리스트 반환
#     return order if len(order) == numCourses else []
